pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1450,65 +1058,194 @@ char *Series_iterGetName(JSOBJ obj, JSONTypeContext *tc, size_t *outLen) {
+ * JT_ARRAY (essentially a sequence we iterate over) whereas the other orients
+ * require a JT_OBJECT context (whereby we extract keys and values from the DataFrame).
+ */
 void DataFrame_iterBegin(JSOBJ obj, JSONTypeContext *tc) {
+    PRINTMARK();
     PyObjectEncoder *enc = (PyObjectEncoder *)tc->encoder;
-    GET_TC(tc)->index = 0;
-    GET_TC(tc)->cStr = PyObject_Malloc(20 * sizeof(char));
-    enc->outputFormat = VALUES;  // for contained series & index
-    if (!GET_TC(tc)->cStr) {
-        PyErr_NoMemory();
-    }
+    enc->originalOutputFormat = enc->outputFormat;
+
+    GET_TC(tc)->index = 0;    
+    // For SPLIT format the index tracks columns->index->data progression
+    if (enc->outputFormat == SPLIT) {
+        PRINTMARK();
+        GET_TC(tc)->cStr = PyObject_Malloc(20 * sizeof(char));
+
+	// TODO: changing the outputFormat was in place when this method only dealt with
+	// the SPLIT orient. Now that this handles quite a few we should probably use
+	// another method for changing the outputFormat of underlying objects (maybe tc->prv)
+        enc->outputFormat = VALUES;  // for contained series & index
+        if (!GET_TC(tc)->cStr) {
+	  PyErr_NoMemory();
+        }
+
+	return;  // always return if modifying outputFormat
+    } else {
+      char *method;
+      if (enc->outputFormat == COLUMNS)
+	method = "items";
+      else
+	method = "iterrows";
+
+      PyObject *iter = PyObject_CallMethod(obj, method, NULL);
+      
+      if (iter == 0) {
+	return;
+      }
+
+      GET_TC(tc)->iterator = iter;
+
+      // The RECORDS format essentially generates a JSON array of Series in the
+      // INDEX format, so set that context during serialization
+      
+      if (enc->outputFormat == RECORDS) {
+	enc->outputFormat = INDEX;
+      }
+
+      return;
+    }      
+
     PRINTMARK();
 }
 
+/* 
+ * Function: DataFrame_iterNext
+ * -----------------------------
+ * Provides instructions how to appropriately iterate the object.
+ *
+ * This is dependent on the orient as mentioned in DataFrame_iterBegin
+ */
 int DataFrame_iterNext(JSOBJ obj, JSONTypeContext *tc) {
-    Py_ssize_t index;
-    if (!GET_TC(tc)->cStr) {
-        return 0;
+    PRINTMARK();
+    PyObjectEncoder *enc = (PyObjectEncoder *)tc->encoder;
+
+    // free previous entry
+    if (GET_TC(tc)->itemValue) {
+      Py_DECREF(GET_TC(tc)->itemValue);
     }
 
-    index = GET_TC(tc)->index;
-    Py_XDECREF(GET_TC(tc)->itemValue);
-    if (index == 0) {
+    // Check the original output format as this may have been modified for
+    // underlying series'
+    if (enc->originalOutputFormat == SPLIT) {
+      Py_ssize_t index;
+      index = GET_TC(tc)->index;
+      if (index == 0) {
         memcpy(GET_TC(tc)->cStr, "columns", sizeof(char) * 8);
+	PRINTMARK();
         GET_TC(tc)->itemValue = PyObject_GetAttrString(obj, "columns");
-    } else if (index == 1) {
+      } else if (index == 1) {
```

Comment:
what are these integers referring?

Reply:
The SPLIT format has custom iteration to write out (in order):

  - 0: the index
  - 1: the columns
  - 2: the data

This was already in place but just wrapped in a conditional statement to have the other orients pass through this context as well. Should certainly refactor to align iteration with other orients (I think better as follow up but will see)

Reply:
k, yeah odd that these are used here (but sure could have been original)